geode-core/src/main/java/org/apache/geode/internal/cache/CacheDistributionAdvisor.java

```diff
@@ -142,17 +149,25 @@ public String toString() {
    *
    * @param excludeInRecovery if true then members in recovery are excluded
    */
-  private Set<InternalDistributedMember> adviseAllEventsOrCached(final boolean excludeInRecovery)
+  private Set<InternalDistributedMember> adviseAllEventsOrCached()
       throws IllegalStateException {
     getAdvisee().getCancelCriterion().checkCancelInProgress(null);
-    return adviseFilter(profile -> {
-      assert profile instanceof CacheProfile;
-      CacheProfile cp = (CacheProfile) profile;
-      if (excludeInRecovery && cp.inRecovery) {
-        return false;
-      }
-      return cp.cachedOrAllEventsWithListener();
-    });
+
+    // minimize volatile reads by copying ref to local var
+    long tempProfilesVersion = profilesVersion; // volatile read
+    if (adviseAllEventsVersion != tempProfilesVersion) {
```

Comment:
I think you may have a java memory model problem with the boolean and Set that maintain the cached state. Neither of these are volatile but the only code that updates them on read a volatile (profilesVersion). You could have two threads that call adviseAllEventOrCached. Since neither of those threads wrote a volatile (profilesVersion) I don't think we have any JMM rules saying we will see the latest values of adviseAllEventsVersion and adviseSetforAllEvents. At first I thought this would be okay if both threads just end up computing the same result and the last one to do so would end up storing its value in the cache. But I think we might have a couple of problems. 1. we are setting two fields and it would be possible for one thread to have a later profilesVersion but then the order they update the two cached fields is not safe so the could win on one and lose on the other. So now the cached state is inconsistent with itself. 2. A thread could see the new boolean and think its adviseSetForAllEvents is up to date but when it reads that field it reads the old stale state. According to my reading of the JMM if you made "adviseAllEventsVersion" also volatile then it would address both of these issues. Another way to code for these issues is to store the cached state in a single AtomicReference whose value is a class with two final fields: a boolean and a set. I think doing it with an AtomicReference is clearer so I would recommend that.

Reply:
I suggested a similar issue and solution in my overall review comment above.

Reply:
Solution updated according to your proposal.

Reply:
Test, are failing, so I will revert this changes

Reply:
The code a currently see looks reasonable (with the additional volatile and added sync). It looks correct. But you still need to add the requests unit and ConcurrentTestRunner tests